<h1 align="center">Hamoye Data Science Internship </h1> 
<h1 align="center"> Stage_A Assessment Codes </h1> 

---

### By  Okwoche, James Ogate
#### (Hamoye Data Science Intern - 2020 Cohort)

---

### Solution to Quiz and Assessment

In [21]:
# Importing the necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import csv
import xlrd # to load the data as excel file incase of lack of internet service

In [12]:
#Load the data (directly from the github link) and read it's content
url = 'https://raw.githubusercontent.com/OkJam/Hamoye-Data-Science-Internship/79a72abea629018c48cb46715cd9655c3ac4a4a4/fuel_ferc1.csv'
Fuel_Data = pd.read_csv(url, error_bad_lines = False) 

Note: Internet is required to load the data from the url. You can also download the data and load it manually.

In [11]:
# print the first 20 rows of the data, to be sure it has correctly loaded it.
Fuel_Data.head(20)

record_id  utility_id_ferc1  report_year  \
0    f1_fuel_1994_12_1_0_7                 1         1994   
1   f1_fuel_1994_12_1_0_10                 1         1994   
2    f1_fuel_1994_12_2_0_1                 2         1994   
3    f1_fuel_1994_12_2_0_7                 2         1994   
4   f1_fuel_1994_12_2_0_10                 2         1994   
5   f1_fuel_1994_12_2_0_13                 2         1994   
6    f1_fuel_1994_12_2_1_1                 2         1994   
7   f1_fuel_1994_12_2_1_10                 2         1994   
8   f1_fuel_1994_12_2_2_10                 2         1994   
9    f1_fuel_1994_12_4_0_1                 4         1994   
10   f1_fuel_1994_12_4_0_2                 4         1994   
11   f1_fuel_1994_12_4_0_4                 4         1994   
12   f1_fuel_1994_12_4_0_5                 4         1994   
13   f1_fuel_1994_12_6_0_1                 6         1994   
14   f1_fuel_1994_12_6_0_2                 6         1994   
15   f1_fuel_1994_12_6_0_4                 6         1994   
16   f1_fuel_1994_12_6_0_7                 6         1994   
17  f1_fuel_1994_12_6_0_10                 6         1994   
18  f1_fuel_1994_12_6_0_13                 6         1994   
19  f1_fuel_1994_12_6_0_14                 6         1994   

        plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0               rockport                coal       ton          5377489   
1   rockport total plant                coal       ton         10486945   
2                 gorgas                coal       ton          2978683   
3                  barry                coal       ton          3739484   
4              chickasaw                 gas       mcf            40533   
5    e. c. gaston-unit 5                coal       ton          2124933   
6       joseph m. farley             nuclear       kgU             2260   
7    resondent's portion                coal       ton          5372133   
8   respondent's portion                coal       ton           817520   
9              units 1-3                coal       ton          1471233   
10             units 1-3                 gas       mcf           262720   
11                unit 4                coal       ton           476779   
12                unit 4                 gas       mcf             6673   
13          clinch river                coal       ton          1655558   
14          clinch river                 oil       bbl             6510   
15       amos-apco share                coal       ton          3520831   
16            amos-total                coal       ton          5327281   
17              glen lyn                coal       ton           666975   
18         kanawha river                coal       ton           391864   
19         kanawha river                 oil       bbl             3505   

    fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0             16.590000                     18.590   
1             16.592000                     18.580   
2             24.130000                     39.720   
3             23.950000                     47.210   
4              1.000000                      2.770   
5             23.922000                     44.240   
6              0.064094                     28.770   
7             24.840000                     58.540   
8             24.290000                     33.970   
9             22.484000                     28.730   
10             1.026000                      1.940   
11            22.428000                     28.710   
12             1.026000                      1.940   
13            25.038000                     30.248   
14             5.875338                     32.130   
15            24.508000                     44.404   
16            24.538000                     44.312   
17            25.836000                     36.268   
18            25.160000                     46.024   
19             5.774958                     33.321   

    fuel_cost_per_unit_d

## **Question 1:** 
Given:  

In [50]:
A = [1,2,3,4,5,6]
B = [113,21,34]
A.extend(B)
print ("The correct answer is A.extend(B) i.e ", A)

The correct answer is A.extend(B) i.e  [1, 2, 3, 4, 5, 6, 113, 21, 34]


## **Question 2:** 
We can create a 3 x 3 identity matrix using:

In [22]:
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## **Question 3:**  
Fuel Cost Per Unit Burned can be generated for each fuel as

In [51]:
Fuel_Data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean().sort_values()

fuel_type_code_pudl
gas          13.659397
other        18.253856
waste        19.518122
coal         67.421830
oil         168.877086
nuclear    4955.157002
Name: fuel_cost_per_unit_burned, dtype: float64

The Fuel with the lowest Fuel Cost Per Unit Burned is **Gas**

## **Question 4:** ### 
Both the standard deviation and the 75th percentile can be obtained from the description of the entire data

In [31]:
Fuel_Data.describe() #The solution is in the appropriate column and rows

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

Or by describing only the column of interest (fuel mmbtu per unit)

In [32]:
Fuel_Data['fuel_mmbtu_per_unit'].describe() # We can also describe only the column of interest. 

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

 **Standard deviation =  10.60022**;
 
 **75th Percentile  =  17.006**

## **Question 5:** 

Skewness OF Fuel Quantity Burned 

In [33]:
Fuel_Data['fuel_qty_burned'].skew()


15.851495469109503

Kurtosis for Fuel Quantity Burned 

In [34]:
Fuel_Data['fuel_qty_burned'].kurt()

651.3694501337732

## **Question 6:**
Checking for missing Data

In [40]:
Fuel_Data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

Feature = fuel_unit

In [56]:
#Total number of data
Total = Fuel_Data['fuel_unit'].count()

In [54]:
# Number of Missing Data 
Missing = Fuel_Data['fuel_unit'].isnull().sum()

In [58]:
# Percentage

Percent = (Missing/Total)*100
print(Percent)


0.6134342091810653


In [ ]:
# using the isnull function to capture the null values and use the sum function to know the total number of the null values
null = fuel_data['fuel_unit'].isnull().sum()

#using the count function on the record id feature to know the total number of rows
total_value = fuel_data['record_id'].count()
total_value1 = fuel_data['fuel_unit'].count()

#using the total rows in the dataset
#checking the percentage of missing values to the total number of rows
percentage = (null/total_value) * 100
percentage

## **Question 7:**
The feature with missing values falls under: 

**Categorical and mode imputation**

## **Question 8:**
The second and third lowest correlation with the Fuel Cost Per Unit Burned are: 

In [62]:
Fuel_Data.corr().loc['fuel_cost_per_unit_burned'].abs().sort_values()

fuel_cost_per_mmbtu             0.000437
fuel_mmbtu_per_unit             0.010034
fuel_cost_per_unit_delivered    0.011007
report_year                     0.013599
fuel_qty_burned                 0.018535
utility_id_ferc1                0.037863
fuel_cost_per_unit_burned       1.000000
Name: fuel_cost_per_unit_burned, dtype: float64

## **Question 9:**

## **Question 10:**
Year with the highest fuel cost per unit delivered

In [61]:
Fuel_Data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean().sort_values(ascending=True)

report_year
1999       25.551627
1995       32.735269
2006       38.657484
2005       41.438184
2007       43.325023
2017       46.196861
2002       47.594361
2003       55.663493
2008       58.588197
2011       59.774667
2001       60.050396
2012       60.994502
1994       63.636060
2010       91.862105
2016      103.901761
2004      139.524275
2013      172.307591
2014      192.737183
1998      287.154420
2015      326.535511
2018      499.269966
2009      652.694163
2000      985.362877
1996     9196.705948
1997    11140.197239
Name: fuel_cost_per_unit_delivered, dtype: float64